# Solución primer reto

### Estudiantes: Maria Katalina Leyva Díaz y Jonatan Palomares Castañeda.


Impostación de librerías:

In [1]:
%pip install matplotlib seaborn plotly scipy streamlit
#Importación de librerías
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import streamlit as st
from scipy import stats
import warnings
warnings.filterwarnings('ignore')


[notice] A new release of pip is available: 25.2 -> 25.3
[notice] To update, run: python.exe -m pip install --upgrade pip


Note: you may need to restart the kernel to use updated packages.


In [22]:
import pandas as pd
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)
pd.set_option('display.max_colwidth', None)

Extraer conjunto de datos del archivo:

In [40]:
df_raw = pd.read_csv("../data/NHE2023.csv", encoding="latin-1", header=None)

# Diagnóstico general de la calidad de los datos:

In [41]:
df_raw.head(20)

0   \
0   NATIONAL HEALTH EXPENDITURES BY TYPE OF SERVICE AND SOURCE OF FUNDS:  CALENDAR YEARS 1960 to 2023   
1                                                                       Expenditure Amount (Millions)   
2                                                                  Total National Health Expenditures   
3                                                                                       Out of pocket   
4                                                                                    Health Insurance   
5                                                                            Private Health Insurance   
6                                                                                            Medicare   
7                                                                                Medicaid (Title XIX)   
8                                                                                           Federal     
9                                                                                     State and Local   
10                                                                     CHIP (Title XIX and Title XXI)   
11                                                                                          Federal     
12                                                                                    State and Local   
13                                                                              Department of Defense   
14                                                                     Department of Veterans Affairs   
15                                                              Other Third Party Payers and Programs   
16                                                                               Worksite Health Care   
17                                                                             Other Private Revenues   
18                                                                            Indian Health Services    
19                                                                              Workers' Compensation   

        1       2       3       4       5       6       7       8       9   \
0      NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN   
1     1960    1961    1962    1963    1964    1965    1966    1967    1968   
2   27,122  29,060  31,765  34,558  38,245  41,627  45,752  51,186  58,015   
3   12,778  13,193  14,085  15,171  16,669  17,863  18,132  17,985  19,951   
4    7,324   8,043   8,785   9,660  10,710  11,722  15,357  20,789  24,010   
5    5,639   6,275   6,963   7,720   8,791   9,772   9,980  10,132  11,462   
6     -       -       -       -       -       -      1,842   4,924   6,218   
7     -       -       -       -       -       -      1,304   3,141   3,541   
8     -       -       -       -       -       -        632   1,525   1,835   
9     -       -       -       -       -       -        672   1,616   1,707   
10    -       -       -       -       -       -       -       -       -      
11    -       -       -       -       -       -       -       -       -      
12    -       -       -       -       -       -       -       -       -      
13     788     836     860     925     861     831   1,078   1,324   1,445   
14     897     933     961   1,015   1,058   1,119   1,153   1,268   1,344   
15   4,081   4,541   4,938   5,461   5,946   6,660   6,563   6,333   7,227   
16      62      66      71      78      87      97     110     128     147   
17     486     542     600     676     774     896   1,000   1,113   1,370   
18    -       -       -       -       -       -       -       -       -      
19     615     644     705     742     807     870     991   1,100   1,244   

        10      11      12      13       14       15       16       17  \
0      NaN     NaN     NaN     NaN      NaN      NaN      NaN      NaN   
1     1969    1970    1971    1972     1973     1974     1975     1976   
2   65,417  74,077  82,380  92,3

Se puede ver que el archivo es transpuesto. Es decir, las variables están ubicadas en la primera posición de cada fila y le siguen sus respectivos datos. Por lo que se realiza su respectiva transformación y se obtiene:

In [42]:
df = df_raw.T
df.columns = df.iloc[0]   #Se convierte la primera fila en encabezados
df = df[1:]
df.head()

0 NATIONAL HEALTH EXPENDITURES BY TYPE OF SERVICE AND SOURCE OF FUNDS:  CALENDAR YEARS 1960 to 2023  \
1                                                                                               NaN   
2                                                                                               NaN   
3                                                                                               NaN   
4                                                                                               NaN   
5                                                                                               NaN   

0 Expenditure Amount (Millions) Total National Health Expenditures  \
1                          1960                             27,122   
2                          1961                             29,060   
3                          1962                             31,765   
4                          1963                             34,558   
5                          1964                             38,245   

0      Out of pocket      Health Insurance           Private Health Insurance  \
1             12,778                 7,324                              5,639   
2             13,193                 8,043                              6,275   
3             14,085                 8,785                              6,963   
4             15,171                 9,660                              7,720   
5             16,669                10,710                              8,791   

0           Medicare           Medicaid (Title XIX)  \
1               -                              -      
2               -                              -      
3               -                              -      
4               -                              -      
5               -                              -      

0                           Federal    \
1                                -      
2                                -      
3                                -      
4                                -      
5                                -      

0                           State and Local  \
1                                      -      
2                                      -      
3                                      -      
4                                      -      
5                                      -      

0           CHIP (Title XIX and Title XXI)  \
1                                     -      
2                                     -      
3                                     -      
4                                     -      
5                                     -      

0                           Federal    \
1                                -      
2                                -      
3                                -      
4                                -      
5                                -      

0                           State and Local          Department of Defense  \
1                                      -                               788   
2                                      -                               836   
3                                      -                               860   
4                                      -                               925   
5                                      -                               861   

0          Department of Veterans Affairs  \
1                                     897   
2                                     933   
3                                     961   
4                                   1,015   
5                                   1,058   

0      Other Third Party Payers and Programs          Worksite Health Care  \
1                                      4,081                            62   
2                                      4,541                            66   
3                                      4,938                            71   
4                                   

In [43]:
df.shape

(64, 545)

En total, se tiene un conjunto de datos de 545 registros y 64 variables.

In [44]:
df.dtypes


0
NATIONAL HEALTH EXPENDITURES BY TYPE OF SERVICE AND SOURCE OF FUNDS:  CALENDAR YEARS 1960 to 2023                                                                                                                                               object
Expenditure Amount (Millions)                                                                                                                                                                                                                   object
Total National Health Expenditures                                                                                                                                                                                                              object
     Out of pocket                                                                                                                                                                                                                              object
     Healt

Como muestra el resultado, la mayoría de columnas están definidas como object o como float64, por lo que será necesaria una transformación para el tratamiento y su análisis durante la limpieza.

In [45]:
# df.isna().sum()
int((df.isna().any(axis=1)).sum())


64

Asimismo, se observa que en las 64 variables hay presencia de valores faltantes.

Dado que hay números definidos con "," con miles y "-" para NaN, se realiza la siguiente transformación para que afecte el procesamiento.

In [46]:
df = df.replace('"', '', regex=True)
df = df.replace('\.', '', regex=True)
df = df.replace(',', '', regex=True)  
df = df.replace('-', None)          

# **Primer reto**

Dado que solo se utilizará la sección de Total National Health Expenditures, se filtrara el conjunto por esta sección y se realizará la limpieza de la misma.

# Filtro de los datos: selección de datos para Total National health Expenditures

In [53]:
names = df.columns.tolist()

start = "Total National Health Expenditures"
end   = "POPULATION"

# encontrar PRIMERA aparición del bloque
i_start = names.index(start)

# encontrar PRIMERA aparición del end DESPUÉS del start
i_end = names.index(end, i_start + 1)

# extraer todas las columnas del bloque (incluye start pero NO end)
block_cols = names[i_start : i_end]

subset = df[block_cols]

subset.head()


0 Total National Health Expenditures       Out of pocket       Out of pocket  \
1                              27122               12778               12778   
2                              29060               13193               13193   
3                              31765               14085               14085   
4                              34558               15171               15171   
5                              38245               16669               16669   

0       Out of pocket       Out of pocket       Out of pocket  \
1               12778                1839                3288   
2               13193                1881                3378   
3               14085                1936                3575   
4               15171                2190                4006   
5               16669                2483                4613   

0       Out of pocket       Out of pocket       Out of pocket  \
1                1906                 340                   7   
2                2018                 342                   8   
3                2135                 350                   8   
4                2275                 359                   8   
5                2515                 381                   9   

0       Out of pocket       Out of pocket       Out of pocket  \
1                1486                2570                 685   
2                1632                2596                 706   
3                1765                2890                 842   
4                1843                3007                 831   
5                1868                3141                 927   

0       Out of pocket       Out of pocket       Out of pocket  \
1                 603                  52                -      
2                 576                  56                -      
3                 524                  59                -      
4                 589                  63                -      
5                 662                  70                -      

0       Out of pocket       Out of pocket       Out of pocket  \
1                 NaN                 NaN                 NaN   
2                 NaN                 NaN                 NaN   
3                 NaN                 NaN                 NaN   
4                 NaN                 NaN                 NaN   
5                 NaN                 NaN                 NaN   

0      Health Insurance      Health Insurance      Health Insurance  \
1                  7324                  7324                  6551   
2                  8043                  8043                  7200   
3                  8785                  8785                  7861   
4                  9660                  9660                  8709   
5                 10710                 10710                  9668   

0      Health Insurance      Health Insurance      Health Insurance  \
1                  4503                  1814                    64   
2                  5011                  1941                    64   
3                  5442                  2146                    66   
4                  5932                  2486                    71   
5                  6464                  2854                    71   

0      Health Insurance      Health Insurance      Health Insurance  \
1                     8                     3                  -      
2                    11                     3                  -      
3                    14                     3                  -      
4                    18                     4                  -      
5                    23                     4                  -      

0      Health Insurance      Health Insurance      Health Insurance  \
1                    37                    40                  -      
2                    40                    40                  -      
3                    44                    49                  -      

In [54]:
subset.shape

(64, 1063)

# Limpieza de datos

Inicialmente, un primer renombramiento de las variables, antes de identificar y tratar valores nulos, atípicos, entre otros.

In [30]:
df.columns

Index(['NATIONAL HEALTH EXPENDITURES BY TYPE OF SERVICE AND SOURCE OF FUNDS:  CALENDAR YEARS 1960 to 2023',
       'Expenditure Amount (Millions)', 'Total National Health Expenditures',
       '     Out of pocket', '     Health Insurance',
       '          Private Health Insurance', '          Medicare',
       '          Medicaid (Title XIX)', '                          Federal  ',
       '                          State and Local',
       ...
       '         Federal Gov't', '         State and Local Gov't',
       '     Equipment', '          Private ', '          Federal Gov't',
       '          State and Local Gov't',
       '* Other Federal Programs include OEO, Federal General and Medical, Federal General and Medical NEC, and High Risk Pools under ACA',
       '** Other State and Local Programs include State and Local Subsidies and TDI',
       'NOTE: Numbers may not add to totals due to rounding. Dollar amounts shown are in current dollars.   "—" Not applicable; Medicare and 

In [32]:
import pandas as pd
pd.set_option('display.max_colwidth', None)
pd.set_option('display.max_rows', None)
pd.set_option('display.max_seq_items', None)

In [34]:
with open("column_names.txt", "w", encoding="utf-8") as f:
    for col in df.columns:
        f.write(col + "\n")



In [ ]:
df = df.rename(columns={
    "OldName1": "NewName1",
    "OldName2": "NewName2"
})


In [29]:
df.describe()

0      NATIONAL HEALTH EXPENDITURES BY TYPE OF SERVICE AND SOURCE OF FUNDS:  CALENDAR YEARS 1960 to 2023  \
count                                                                                                0.0   
mean                                                                                                 NaN   
std                                                                                                  NaN   
min                                                                                                  NaN   
25%                                                                                                  NaN   
50%                                                                                                  NaN   
75%                                                                                                  NaN   
max                                                                                                  NaN   

0           Out of pocket            Private Health Insurance  \
count                 0.0                                 0.0   
mean                  NaN                                 NaN   
std                   NaN                                 NaN   
min                   NaN                                 NaN   
25%                   NaN                                 NaN   
50%                   NaN                                 NaN   
75%                   NaN                                 NaN   
max                   NaN                                 NaN   

0                Medicare                            Federal    \
count                 0.0                                  0.0   
mean                  NaN                                  NaN   
std                   NaN                                  NaN   
min                   NaN                                  NaN   
25%                   NaN                                  NaN   
50%                   NaN                                  NaN   
75%                   NaN                                  NaN   
max                   NaN                                  NaN   

0                                Federal             Department of Defense  \
count                                  0.0                             0.0   
mean                                   NaN                             NaN   
std                                    NaN                             NaN   
min                                    NaN                             NaN   
25%                                    NaN                             NaN   
50%                                    NaN                             NaN   
75%                                    NaN                             NaN   
max                                    NaN                             NaN   

0               Department of Veterans Affairs           Worksite Health Care  \
count                                      0.0                            0.0   
mean                                       NaN                            NaN   
std                                        NaN                            NaN   
min                                        NaN                            NaN   
25%                                        NaN                            NaN   
50%                                        NaN                            NaN   
75%                                        NaN                            NaN   
max                                        NaN                            NaN   

0               Other Private Revenues           Indian Health Services   \
count                              0.0                               0.0   
mean                               NaN                               NaN   
std                                NaN                               NaN   
min                                NaN                               NaN   
25%                              

# **Primer reto**

### 1. Subconjunto de dataframe 